File: others/permutate.py
def permutate(s: str) -> list[str]:
    if(len(s) == 1):
        return [s]
    
    res = []

    for i in range(len(s)):
        char = s[i]
        rest = s[:i] + s[i + 1:]

        perms = permutate(rest)

        for perm in perms:
            res.append(char + perm)

    return res


result = permutate('abc')
print(result)

File: others/permutate.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
const permutate = (str) => {
  let res = [];

  if (str.length === 1) return [str];

  for (let i = 0; i < str.length; i++) {
    let char = str[i];
    let rest = str.substring(0, i) + str.substring(i + 1);

    let perms = permutate(rest);

    for (const perm of perms) {
      res.push(char + perm);
    }
  }

  return res;
};

console.log(permutate("abc"));
